cupyx/scipy/sparse/linalg/solve.py

```diff
@@ -65,3 +66,266 @@ def lsqr(A, b):
     x = x.astype(numpy.float64)
     ret = (x, None, None, None, None, None, None, None, None, None)
     return ret
+
+
+def bicgstab(A, b, x0=None, M=None, tol=1e-5, maxiter=numpy.inf,
+             callback=None, atol='legacy'):
+    """Solves linear system using an iterative solver BiCGSTAB.
+
+    See https://docs.nvidia.com/cuda/cusparse/#csrilu02_solve and
+    https://docs.nvidia.com/cuda/incomplete-lu-cholesky/index.html.
+
+    Find the solution to a large, sparse, linear system of equations.
+    The function approximates ``Ax = b`` using BiCGSTAB iterative method.
+
+    Args:
+        A (cupy.ndarray or cupyx.scipy.sparse.csr_matrix): The input matrix
+            with dimension ``(N, N)``
+        b (cupy.ndarray): Right-hand side vector.
+        x0 (cupy.ndarray): Initial guess (defaults to zeros)
+        M (tuple(callable, optional(info))): GPU preconditioner function
+            and info object for preconditioner (to destroy after running)
```

Comment:
I don't understand how to make this work. I think we'll need
1. A docstring `.. notes::` providing an example for nontrivial `M`
2. Test coverage (currently lacking)

Reply:
Yes, I agree I should add some tests for this. The `spilu` preconditioner is actually one example. I can write a docstring with a spilu example and make a test for that. Would that work for you?

Reply:
Sounds good. This would then becomes an integration test, which is also good. Are they other possible preconditioners either existing in the codebase or can be made up quickly?

Reply:
The Jacobi one is simple as well but it requires implementing `A.diagonal()` which I think is not yet implemented.